### PR TITLE
Avoid re-jitting distributed tasks in the same process

### PR DESCRIPTION
**Pull request by vepadulano**

### CODE REVIEW

**Comment by vepadulano** [comment]
## Example benchmark of the effects of this PR

### Setup
Hardware

* AMD 5950x (16 physical cores)
* 64 GB RAM
* 1 TB NVMe SSD

Software

* Python 3.11, GCC 12
* dask 2023.2.0

Benchmark

* dimuon tutorial, using distrdf+dask
* 1 file, stored on local SSD. The file is preloaded in filesystem cache before the tests run
* ~2GB of data, 75 TTree clusters
* 2 physical cores. One Dask worker is spawned on each core

The benchmark is run keeping the number of cores fixed, increasing the number of partitions per core. Each test is run 10 times, plots show mean value at each number of partitions and the error is the standard error of the mean.

## Results
### Memory usage

Memory is taken from `psutil.virtual_memory().available`, i.e. the total available memory on the system. The benchmark is the "only" process running on the system (i.e. the machine is disconnected from the internet and no other user applications are running at the same time).

Memory used by the test run is computed as the difference between the memory available before running the analysis and the memory available at the end of the analysis, before exiting the Python script.

avoid_rejitting_memory

### Runtime

Runtime is computed as the runtime of the analysis, from right before creating the RDataFrame to the moment when the results of the computations are obtained.

avoid_rejitting_runtime

File: bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py [hunk 1]
```diff
@@ -241,3 +241,34 @@ def _(mergeable: ROOT.Detail.RDF.RMergeableVariationsBase, node, backend):
...
+def _(asnumpyres):
+    asnumpyres._result_ptrs = {col: ROOT.Internal.RDF.CloneResultPtr(ptr) for (col, ptr) in asnumpyres._result_ptrs.items()}
+    return asnumpyres
+
+
+def clone_snapshotresult(snap: SnapshotResult, range_id: int):
+    """
+    Clone the action held by an RResultPtr, registering it with its RLoopManager
+    """
+    name_with_old_id = snap.filenames[0].removesuffix(".root")
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py]: what happens if the output file name does not end in `.root`? If it's required that it ends in `.root`, are we checking for that anywhere?

File: bindings/experimental/distrdf/python/DistRDF/Ranges.py [hunk 2]
```diff
@@ -14,14 +15,17 @@
...
 
+@dataclass
 class DataRange:
     """
     A logical range of entries in which a dataset is split. Depending on the
     input data source, this can have different attributes.
     """
-    pass
+    rdf_uuid: uuid.UUID
+    id: int
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/Ranges.py:25]: need some docs here about what these are and what purpose they serve

File: bindings/experimental/distrdf/python/DistRDF/Ranges.py [hunk 3]
```diff
@@ -39,14 +43,11 @@ class EmptySourceRange(DataRange):
     end (int): Ending entry of this range.
     """
 
-    def __init__(self, rangeid, start, end):
-        """set attributes"""
-        self.id = rangeid
-        self.start = start
-        self.end = end
+    start: int
+    end: int
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/Ranges.py:47]: there is still docs about `id` in this class but the data member was moved to `DataRange`

File: tree/dataframe/src/RResultPtr.cxx [hunk 4]
```diff
@@ -14,15 +14,15 @@
 #include <ROOT/RDF/RInterface.hxx>
 #include <ROOT/RDF/RLoopManager.hxx>
 
-using SnapshotPtr_t = ROOT::RDF::RResultPtr<ROOT::RDF::RInterface<ROOT::Detail::RDF::RLoopManager, void>>;
```
Comment (eguiraud) [inline tree/dataframe/src/RResultPtr.cxx]: Commit messages are to explain the why (and for complex changes the intricacies of the how) that might be lost in the code. It's not clear to me why this change was needed, e.g. if it's just to avoid duplication:

```
[DF] Clean up: remove duplicate type alias declaration
```

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 5]
```diff
@@ -1647,6 +1647,24 @@ public:
...
+   SnapshotHelper MakeNew(void *newName)
+   {
+      // Create local vector of booleans for fIsDefine
+      std::vector<bool> localIsDefine = [&] {
+         std::vector<bool> isDef;
+         isDef.reserve(sizeof...(ColTypes));
+         for (std::size_t i = 0u; i < sizeof...(ColTypes); ++i)
+            isDef[i] = fIsDefine[i];
+         return isDef;
+      }();
```
[the remark this answers is not in the document]
Comment (vepadulano) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx]: I agree that it shouldn't be needed. I originally had some issue when compiling without this trick. It seems fine now, I don't know what was happening previously

File: tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx [hunk 6]
```diff
@@ -1647,6 +1647,24 @@ public:
    {
       return [this](unsigned int, const RSampleInfo &) mutable { fBranchAddressesNeedReset = true; };
    }
+
+   SnapshotHelper MakeNew(void *newName)
```
[the remark this answers is not in the document]
Comment (vepadulano) [reply tree/dataframe/inc/ROOT/RDF/ActionHelpers.hxx:1662]: Split this change in 2 commits in the most recent update. Also, implemented the `MakeNew` method for `SnapshotHelperMT`

File: tree/dataframe/inc/ROOT/RDF/RResultMap.hxx [hunk 7]
```diff
@@ -52,6 +52,35 @@ MakeResultMap(std::shared_ptr<T> nominalResult, std::vector<std::shared_ptr<T>>
    return ROOT::RDF::Experimental::RResultMap<T>(std::move(nominalResult), std::move(variedResults), std::move(keys),
                                                  lm, std::move(nominalAction), std::move(variedAction));
 }
+
+/**
+ * \brief Clones the varied action connected to an RResultMap.
+ *
+ * \tparam T The type of the results held by the map.
+ * \param inmap The map.
+ * \return ROOT::RDF::Experimental::RResultMap<T> A new map with the clone variations.
```
[the remark this answers is not in the document]
Comment (vepadulano) [reply tree/dataframe/inc/ROOT/RDF/RResultMap.hxx]: Too much star wars probably

File: tree/dataframe/inc/ROOT/RResultPtr.hxx [hunk 8]
```diff
@@ -41,7 +41,35 @@ class RInterface;
...
+ * \param inptr The pointer.
+ * \return A new pointer with a cloned action.
+ */
+template <typename T>
+ROOT::RDF::RResultPtr<T> CloneResultPtr(const ROOT::RDF::RResultPtr<T> &inptr)
+{
+   // For now this is an extra copy, best would be calling default constructor.
+   // Main reason (as far as I am aware) is copying metadata of certain
+   // result types, e.g. a for a TH1D we have to create a new histogram with
+   // the same binning and axis limits.
```
[the remark this answers is not in the document]
Comment (vepadulano) [reply tree/dataframe/inc/ROOT/RResultPtr.hxx]: Fair enough, reworded!

File: tree/dataframe/src/RResultPtr.cxx [hunk 9]
```diff
@@ -0,0 +1,30 @@
...
+ *
+ * \param inptr The pointer.
+ * \param newNameForSnapshot A new name for the output file of the cloned action.
+ * \return A new pointer with a cloned action.
+ */
+SnapshotPtr_t ROOT::Internal::RDF::CloneResultPtr(const SnapshotPtr_t &inptr, const std::string &newNameForSnapshot)
+{
+   return SnapshotPtr_t{
+      inptr.fObjPtr, inptr.fLoopManager,
+      inptr.fActionPtr->CloneAction(reinterpret_cast<void *>(const_cast<std::string *>(&newNameForSnapshot)))};
```
[the remark this answers is not in the document]
Comment (vepadulano) [reply tree/dataframe/src/RResultPtr.cxx]: Not really unfortunately, cast to `void *` discards qualifier and the compiler complains, I need the `const_cast` (or a cast to `const void *`, but that would clash with the signature of `MakeNew(void *)`)

File: bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py (hunk 1 above)
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py]: Indeed it should not be an issue, I extended the logic to also accomodate files that do not end in `.root`

File: bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py [hunk 10]
```diff
@@ -241,3 +241,34 @@ def _(mergeable: ROOT.Detail.RDF.RMergeableVariationsBase, node, backend):
...
+@singledispatch
+def clone_action(resultptr):
+    """
+    Clone the action held by an RResultPtr, registering it with its RLoopManager
+    """
+    return ROOT.Internal.RDF.CloneResultPtr(resultptr)
+
+
+@clone_action.register(AsNumpyResult)
+def _(asnumpyres):
```
[the remark this answers is not in the document]
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/Backends/Utils.py]: In the latest update, the logic is implemented fully via `singledispatch`

**Comment by vepadulano** [comment]
Thank you so much @eguiraud for the comments! I have addressed all of them and updated the commits accordingly, also splitting them in smaller commits to make intentions clearer.

New tests are still missing, but I would appreciate if you could take another look. In particular, these changes hopefully streamline the code a bit:
* `rdf_uuid`: this was previously not well named, since we need something that contains both an identifier for the current RDataFrame instance and also one for the current (distributed) execution of that RDataFrame (e.g. to separate to consecutive triggers of the computation graph). I have added the type `ExecutionIdentifier` that holds both. This class is now used everywhere (in Ranges.py, HeadNode.py) and should improve readability.
* I merged `CloneResultPtr` and `CloneResultMap` into `CloneResultAndAction`. This helps making the logic in DistRDF more simple (i.e. is what allows the usage of `singledispatch` to distinguish different ways of cloning actions of the distributed computation graph).

File: tree/dataframe/src/RResultPtr.cxx (hunk 9 above)
[the remark this answers is not in the document]
Comment (eguiraud) [reply tree/dataframe/src/RResultPtr.cxx]: ah right, then only the other side can be simplified: you can directly cast `void *` to `const std::string *`

**Comment by eguiraud** [comment]
Thank you @vepadulano !

I'm happy to do another pass, but I'd rather do so after the tests are added: it takes me around half a day of work to go through a PR this size, I'd like to reduce the amount of passes... :sweat_smile: 

(the warnings are real and the test failure is because of an expected error message that needs to be updated in the test -- or the new message needs to match the old one)

**Comment by vepadulano** [comment]
This now depends on #12363 for the failures on ubuntu18

**Comment by vepadulano** [comment]
Added tests for the C++ mechanisms, and attempt at fixing the ubuntu18 failure with a much more simplified approach w.r.t. https://github.com/root-project/root/pull/12981, which should not be needed anymore (let's see what the CI thinks).

**Comment by vepadulano** [comment]
@phsft-bot build

**Comment by vepadulano** [comment]
Added tests on the distributed RDF side too.

@eguiraud this is now ready for review again

File: bindings/experimental/distrdf/python/DistRDF/PythonMergeables.py [hunk 11]
```diff
@@ -48,3 +49,7 @@ def GetValue(self, backend: BaseBackend):
             snapshot_chain.Add(filename)
         # Create a new rdf with the chain and return that to user
         return backend.make_dataframe(snapshot_chain)
+
+
+# A type alias to signify any type of result that can be returned from the RDataFrame API
+RDataFrameFutureResult = Union[ROOT.RDF.RResultPtr, ROOT.RDF.Experimental.RResultMap, SnapshotResult, AsNumpyResult]
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/PythonMergeables.py:59]: not for this PR, but maybe `AsNumpyResult` should live in this module and be imported by `_pythonization._rdataframe` instead?

File: bindings/experimental/distrdf/python/DistRDF/_graph_cache.py [hunk 12]
```diff
@@ -0,0 +1,30 @@
+from __future__ import annotations
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/_graph_cache.py:1]: I guess we don't need this, if we are using dataclasses we assume at least Python 3.7

File: bindings/experimental/distrdf/python/DistRDF/_graph_cache.py [hunk 13]
```diff
@@ -0,0 +1,30 @@
...
+
+import uuid
+from typing import Dict, TYPE_CHECKING
+
+if TYPE_CHECKING:
+    import ROOT
+    from DistRDF.PythonMergeables import RDataFrameFutureResult
+
+
+@dataclass(eq=True, frozen=True)
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/_graph_cache.py:12]: isn't `eq=True` the default?

File: bindings/experimental/distrdf/python/DistRDF/Ranges.py [hunk 14]
```diff
@@ -157,6 +128,51 @@ class TaskTreeEntries:
     trees_with_entries: Dict[str, int] = field(default_factory=dict)
 
 
+def get_balanced_ranges(nentries, npartitions, exec_id: ExecutionIdentifier):
+    """
+    Builds range pairs from the given values of the number of entries in
+    the dataset and number of partitions required. Each range contains the
+    same amount of entries, except for those cases where the number of
+    entries is not a multiple of the partitions.
```
Comment (eguiraud) [inline bindings/experimental/distrdf/python/DistRDF/Ranges.py:136]: maybe more simply you can just say "The `nentries` are divided uniformly among the `npartitions`".

File: tree/dataframe/test/dataframe_cloning.cxx [hunk 15]
```diff
@@ -0,0 +1,405 @@
...
+   for (unsigned i = 1; i < 6; i++) {
+      ChangeSpec(df, std::move(specs[i]));
+      auto clone = CloneResultAndAction(take);
+      EXPECT_VEC_EQ(*clone, expectedOutputs[i]);
+   }
+
+   for (const auto &name : fileNames) {
+      gSystem->Unlink(name.c_str());
+   }
+}
```
Comment (eguiraud) [inline tree/dataframe/test/dataframe_cloning.cxx:405]: NICE

**Review by eguiraud — APPROVED** [review]
LGTM! Test coverage looks great and the new commit splitting makes a lot of sense.

Only minor comments, feel free to ignore.

File: tree/dataframe/src/RLoopManager.cxx [hunk 16]
```diff
@@ -369,20 +369,31 @@ RLoopManager::RLoopManager(std::unique_ptr<RDataSource> ds, const ColumnNames_t
...
+   : fNSlots(RDFInternal::GetNSlots()),
      fLoopType(ROOT::IsImplicitMTEnabled() ? ELoopType::kROOTFilesMT : ELoopType::kROOTFiles),
      fNewSampleNotifier(fNSlots),
      fSampleInfos(fNSlots),
      fDatasetColumnReaders(fNSlots)
 {
+   ChangeSpec(std::move(spec));
+}
+
+void RLoopManager::ChangeSpec(ROOT::RDF::Experimental::RDatasetSpec &&spec)
```
[the remark this answers is not in the document]
Comment (vepadulano) [reply tree/dataframe/src/RLoopManager.cxx:381]: Indeed, the new spec is expected to provide exactly the same dataset schema. I agree with you that there is a very dangerous interaction potentially. We can't really do otherwise for now. I will add some very explicit docs in the next commit

File: bindings/experimental/distrdf/python/DistRDF/PythonMergeables.py (hunk 11 above)
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/PythonMergeables.py:59]: Unclear, `AsNumpyResult` is actually tied to the `AsNumpy` feature, before being used in `DistRDF`. Moving it here would mean making that part of PyROOT depend on a class defined in a `DistRDF` submodule. One alternative is splitting the class in two. The actual result of `AsNumpy` would still live only in PyROOT, whereas in `DistRDF` I could create a second class and call it `AsNumpyFuture` or similar, that wraps the `AsNumpyResult` from PyROOT and then adds the `Merge` functionality on top

File: bindings/experimental/distrdf/python/DistRDF/_graph_cache.py (hunk 12 above)
Comment (vepadulano) [reply bindings/experimental/distrdf/python/DistRDF/_graph_cache.py:1]: Delayed evaluation of annotations is a Python 3.7 feature indeed, but needs to be enforced via the `__future__` import. Unfortunately, the Python standard hasn't yet decided on whether this will actually be the default behaviour (see https://docs.python.org/3/library/__future__.html#id1). In DistRDF we use it extensively both to avoid circular imports and to avoid importing modules that are used only for annotations, so I just write the `__future__` import in every module

File: tree/dataframe/inc/ROOT/RDF/RVariedAction.hxx [hunk 17]
```diff
@@ -194,6 +194,22 @@ public:
       throw std::logic_error("Cannot produce a varied action from a varied action.");
    }
 
+   std::unique_ptr<RActionBase> CloneAction(void *typeErasedResults) final
+   {
+      const auto &vectorOfTypeErasedResults = *reinterpret_cast<const std::vector<void *> *>(typeErasedResults);
```
[the remark this answers is not in the document]
Comment (vepadulano) [reply tree/dataframe/inc/ROOT/RDF/RVariedAction.hxx:199]: I don't think that's possible since `RActionBase` declares
```
virtual std::unique_ptr<RActionBase> CloneAction(void *newResult) = 0;
```

[the remark this answers is not in the document]
Comment (vepadulano) [reply tree/dataframe/inc/ROOT/RDF/RVariedAction.hxx:199]: I mean, I could always make a `RVariedAction::CloneActionImpl(const std::vector<void *> &results)` and then `RVariedAction::CloneAction(void *results)` would just reinterpret cast the pointer to a vector and call the `Impl` function, but that's not really necessary imho
